File: Persistencia/Ejercicio 1/python/archivo_empleado.py
import json
from typing import Optional, List

class Empleado:
    def __init__(self, nombre: str, edad: int, salario: float):
        self.nombre = nombre
        self.edad = edad
        self.salario = salario
    
    def to_dict(self):
        return {"nombre": self.nombre, "edad": self.edad, "salario": self.salario}

class ArchivoEmpleado:
    def __init__(self, nombre: str):
        self.nombre = nombre
    
    def crearArchivo(self) -> None:
        try:
            with open(self.nombre, 'w') as f:
                json.dump([], f)
        except IOError:
            print("Error, el archivo no fue creadod.")
    
    def guardarEmpleado(self, e: Empleado) -> None:
        try:
            with open(self.nombre, 'r') as f:
                empleados = json.load(f)
        except (IOError, json.JSONDecodeError):
            empleados = []
        empleados.append(e.to_dict())
        try:
            with open(self.nombre, 'w') as f:
                json.dump(empleados, f, indent=4)
        except IOError:
            print("Error, no se guardo el empleado.")
    
    def buscaEmpleado(self, n: str) -> Optional[Empleado]:
        try:
            with open(self.nombre, 'r') as f:
                empleados = json.load(f)
                for e in empleados:
                    if e["nombre"] == n:
                        return Empleado(e["nombre"], e["edad"], e["salario"])
        except (IOError, json.JSONDecodeError):
            print("Error, no se busco el empleado.")
        return None
    
    def mayorSalario(self, sueldo: float) -> Optional[Empleado]:
        try:
            with open(self.nombre, 'r') as f:
                empleados = json.load(f)
                mayor = None
                for e in empleados:
                    if e["salario"] > sueldo:
                        if mayor is None or e["salario"] > mayor.salario:
                            mayor = Empleado(e["nombre"], e["edad"], e["salario"])
                return mayor
        except (IOError, json.JSONDecodeError):
            print("Error, no se busco al mayor salario.")
        return None


archivo = ArchivoEmpleado("empleados.json")
archivo.crearArchivo()
archivo.guardarEmpleado(Empleado("Diego", 20, 4000.0))
archivo.guardarEmpleado(Empleado("Antonio", 35, 6000.0))
encontrado = archivo.buscaEmpleado("Diego")
if encontrado: print("Encontrado:", encontrado.nombre, encontrado.edad, encontrado.salario)
mayor = archivo.mayorSalario(5000.0)
if mayor: print("Mayor salario:", mayor.nombre, mayor.edad, mayor.salario)

File: Persistencia/Ejercicio 1/java/ArchivoEmpleado.java
import java.io.*;
import java.util.ArrayList;

public class ArchivoEmpleado {
    private String nombre;
    
    public ArchivoEmpleado(String nombre) {
        this.nombre = nombre;
    }
    
    public void crearArchivo() {
        try {
            new FileOutputStream(nombre).close();
        } catch (IOException e) {
            System.out.println("Error, no se creo el archivo.");
        }
    }
    
    public void guardarEmpleado(Empleado e) {
        try {
            FileOutputStream file = new FileOutputStream(nombre, true);
            ObjectOutputStream out;
            if (new File(nombre).length() == 0) {
                out = new ObjectOutputStream(file);
            } else {
                out = new ObjectOutputStream(file) {
                    protected void writeStreamHeader() throws IOException {
                    }
                };
            }
            out.writeObject(e);
            out.close();
            file.close();
        } catch (IOException ex) {
            System.out.println("Error, no se guardo empleado.");
        }
    }
    
    public Empleado buscaEmpleado(String n) {
        FileInputStream file = null;
        ObjectInputStream in = null;
        try {
            file = new FileInputStream(nombre);
            in = new ObjectInputStream(file);
            while (true) {
                Empleado e = (Empleado) in.readObject();
                if (e.nombre.equals(n)) {
                    return e;
                }
            }
        } catch (EOFException e) {
            return null;
        } catch (IOException | ClassNotFoundException ex) {
            System.out.println("Error, no se busco empleado.");
            return null;
        } finally {
            try {
                if (in != null) in.close();
                if (file != null) file.close();
            } catch (IOException e) {
                System.out.println("Error al cerrar el archivo.");
            }
        }
    }
    
    public Empleado mayorSalario(float sueldo) {
        Empleado mayor = null;
        FileInputStream file = null;
        ObjectInputStream in = null;
        try {
            file = new FileInputStream(nombre);
            in = new ObjectInputStream(file);
            ArrayList<Empleado> empleados = new ArrayList<>();
            while (true) {
                try {
                    Empleado e = (Empleado) in.readObject();
                    empleados.add(e);
                } catch (EOFException e) {
                    break; 
                }
            }
            for (Empleado e : empleados) {
                if (e.salario > sueldo) {
                    if (mayor == null || e.salario > mayor.salario) {
                        mayor = e;
                    }
                }
            }
        } catch (IOException ex) {
            System.out.println("Error, no se busco mayor salario: " + ex.getMessage());
        } catch (ClassNotFoundException ex) {
            System.out.println("Error no se encontro: " + ex.getMessage());
        } finally {
            try {
                if (in != null) in.close();
                if (file != null) file.close();
            } catch (IOException e) {
                System.out.println("Error al cerrar el archivo.");
            }
        }
        return mayor;
    }
}